### PR TITLE
#3296 Typing Notification update

### PR DESCRIPTION
**Pull request by ahmed-radhouane**
## Issue #3296 Typing Notification update

Signed-off-by: Ahmed Radhouane Belkilani <arbelkilani@gmail.com>

 
## Type of change

- [ ] Feature
- [x] Bugfix
- [ ] Technical
- [ ] Other :

## Content

 - While user is in a room a typing notification appear in the toolbar with the sender name and a short message "user A is typing..." 
 - This will be removed with the issues #4642
 - In this implementation we added a new typing notification view at the bottom of the time line showing some loading dots animation and sending users avatars with new text to appear related to different states (one, two or more senders) . please refer to the Gifs. 
 
## Motivation and context

- A new string resources has been added to the strings.xml file with a //TODO tag above. 
- please refer to the [Figma](https://www.figma.com/file/uaWc62Ux2DkZC4OGtAGcNc/Voice-Messages?node-id=1855%3A86441) file describing the new UI for the typing notification. 

## Screenshots / GIFs

On sender | Two sender
--- | --- 
device-2022-02-11-004242_1 | device-2022-02-11-003851_1

Multi-senders | RTL 
--- | --- 
device-2022-02-11-003404_1 | device-2022-02-11-112149_20220211112314_1
## Tests

- Get into a room from user A connected with User B 
- Type a message in room as User A
- Access to same room as User B ans see the typing notification at the bottom of the timeline. 

## Tested devices

- [x] Physical
- [x] Emulator
- OS version(s): 

## Checklist

- [x] Changes has been tested on an Android device or Android emulator with API 21
- [x] UI change has been tested on both light and dark themes
- [x] Accessibility has been taken into account. See https://github.com/vector-im/element-android/blob/develop/CONTRIBUTING.md#accessibility
- [x] Pull request is based on the develop branch
- [x] Pull request includes a new file under ./changelog.d. See https://github.com/vector-im/element-android/blob/develop/CONTRIBUTING.md#changelog
- [x] Pull request includes screenshots or videos if containing UI changes
- [x] Pull request includes a [sign off](https://matrix-org.github.io/synapse/latest/development/contributing_guide.html#sign-off)
- [x] You've made a self review of your PR
- [ ] If you have modified the screen flow, or added new screens to the application, you have updated the test [UiAllScreensSanityTest.allScreensTest()](https://github.com/vector-im/element-android/blob/main/vector/src/androidTest/java/im/vector/app/ui/UiAllScreensSanityTest.kt#L73)

### CODE REVIEW

File: .idea/shelf/Uncommitted_changes_before_rebase__Default_Changelist_.xml [hunk 1]
```diff
@@ -0,0 +1,4 @@
+<changelist name="Uncommitted_changes_before_rebase_[Default_Changelist]" date="1644570474870" recycled="true" deleted="true">
```
Comment (ouchadam) [inline .idea/shelf/Uncommitted_changes_before_rebase__Default_Changelist_.xml:1]: looks like a bonus file was included!

File: vector/src/main/java/im/vector/app/core/ui/views/TypingMessageAvatar.kt [hunk 2]
```diff
@@ -0,0 +1,60 @@
...
+        attrs: AttributeSet? = null,
+        defStyleAttr: Int = 0
+) : LinearLayout(context, attrs, defStyleAttr) {
+
+    companion object {
+        const val AVATAR_SIZE_DP = 20
+        const val OVERLAP_FACT0R = -3 // =~ 30% to left
+    }
+
+    fun render(typingUsers: List<SenderInfo>, avatarRender: AvatarRenderer) {
```
Comment (ouchadam) [inline vector/src/main/java/im/vector/app/core/ui/views/TypingMessageAvatar.kt:41]: have you noticed any performance issues with creating/dropping views as part of the typing flow?

I'm wondering if it's worth caching and reusing the imageviews instead of dynamically creating/removing them, probably not~

File: .idea/shelf/Uncommitted_changes_before_rebase__Default_Changelist_.xml (hunk 1 above)
Comment (ahmed-radhouane) [reply .idea/shelf/Uncommitted_changes_before_rebase__Default_Changelist_.xml]: it run away, catch it !

File: vector/src/main/java/im/vector/app/core/ui/views/TypingMessageDotsView.kt [hunk 3]
```diff
@@ -0,0 +1,92 @@
...
+        super.onAttachedToWindow()
+        circles.forEachIndexed { index, circle -> animateCircle(index, circle) }
+    }
+
+    private fun animateCircle(index: Int, circle: View) {
+        val animator = ValueAnimator.ofFloat(DEFAULT_MAX_ALPHA, DEFAULT_MIN_ALPHA)
+        animator.duration = DEFAULT_CIRCLE_DURATION
+        animator.startDelay = DEFAULT_START_ANIM_CIRCLE_DURATION * index
+        animator.repeatCount = ValueAnimator.INFINITE
+        animator.addUpdateListener {
```
Comment (ouchadam) [inline vector/src/main/java/im/vector/app/core/ui/views/TypingMessageDotsView.kt:82]: could also use `ObjectAnimator` to avoid needing the separate callback, I don't think there's any difference other than syntax~

```kotlin
val animator = ObjectAnimator.ofFloat(circle, "alpha", DEFAULT_MAX_ALPHA, DEFAULT_MIN_ALPHA)
```

File: vector/src/main/res/layout/vector_settings_round_avatar.xml [hunk 4]
```diff
@@ -1,8 +1,8 @@
 <?xml version="1.0" encoding="utf-8"?>
 <FrameLayout xmlns:android="http://schemas.android.com/apk/res/android"
     xmlns:tools="http://schemas.android.com/tools"
-    android:layout_width="40dp"
-    android:layout_height="40dp">
+    android:layout_width="24dp"
```
Comment (ouchadam) [inline vector/src/main/res/layout/vector_settings_round_avatar.xml:4]: is this layout used anywhere else? this change might effect another screen~

**Review by ouchadam — COMMENTED** [review]
great work! the change looks awesome :100:  

added a few minor comments/queries

File: vector/src/main/res/layout/vector_settings_round_avatar.xml (hunk 4 above)
Comment (ahmed-radhouane) [reply vector/src/main/res/layout/vector_settings_round_avatar.xml]: It was an error, I've revert value to those of develop branch. 
Thanks

File: vector/src/main/java/im/vector/app/core/ui/views/TypingMessageDotsView.kt (hunk 3 above)
Comment (ahmed-radhouane) [reply vector/src/main/java/im/vector/app/core/ui/views/TypingMessageDotsView.kt]: Changed ! 
Thanks

File: vector/src/main/java/im/vector/app/core/ui/views/TypingMessageAvatar.kt (hunk 2 above)
Comment (ahmed-radhouane) [reply vector/src/main/java/im/vector/app/core/ui/views/TypingMessageAvatar.kt:41]: I do not think we need to handle cache at this level. 
avatars count are limit to 4, and users could change rapidly in a short time. 
in total the view took almost 1 second and disappear. Still I will try to renforce test on this. 
Thank you

**Review by ouchadam — APPROVED** [review]
LGTM :100:

File: matrix-sdk-android/src/main/java/org/matrix/android/sdk/api/session/room/sender/SenderInfo.kt [hunk 5]
```diff
@@ -35,3 +36,5 @@ data class SenderInfo(
             else                                      -> "$displayName ($userId)"
         }
 }
+
+fun SenderInfo.toMatrixItem() = MatrixItem.UserItem(userId, displayName, avatarUrl)
```
Comment (bmarty) [inline matrix-sdk-android/src/main/java/org/matrix/android/sdk/api/session/room/sender/SenderInfo.kt:40]: All extensions like this are declared in the file `MatrixItem.kt`.  Maybe move this one there for coherency.

File: vector/src/main/java/im/vector/app/core/ui/views/TypingMessageView.kt [hunk 6]
```diff
@@ -0,0 +1,55 @@
...
+
+    @Inject
+    lateinit var typingHelper: TypingHelper
+
+    init {
+        inflate(context, R.layout.typing_message_layout, this)
+        views = TypingMessageLayoutBinding.bind(this)
+    }
+
+    fun render(typingUsers: List<SenderInfo>, avatarRender: AvatarRenderer) {
```
Comment (bmarty) [inline vector/src/main/java/im/vector/app/core/ui/views/TypingMessageView.kt:46]:
```suggestion
    fun render(typingUsers: List<SenderInfo>, avatarRenderer: AvatarRenderer) {
```

File: vector/src/main/java/im/vector/app/core/ui/views/TypingMessageView.kt [hunk 7]
```diff
@@ -0,0 +1,55 @@
...
+        views.usersName.text = typingHelper.getNotificationTypingMessage(typingUsers)
+        views.avatars.render(typingUsers, avatarRender)
```
Comment (bmarty) [inline vector/src/main/java/im/vector/app/core/ui/views/TypingMessageView.kt:48]:
```suggestion
        views.avatars.render(typingUsers, avatarRenderer)
```

File: vector/src/main/java/im/vector/app/features/home/room/detail/TimelineFragment.kt [hunk 8]
```diff
@@ -1597,6 +1601,17 @@ class TimelineFragment @Inject constructor(
         voiceMessageRecorderView.isVisible = false
     }
 
+    private fun renderTypingMessageNotification(roomSummary: RoomSummary?, state: RoomDetailViewState) {
+        if (!isThreadTimeLine() && roomSummary != null) {
+            views.typingMessageView.isInvisible = state.typingUsers.isNullOrEmpty()
```
Comment (bmarty) [inline vector/src/main/java/im/vector/app/features/home/room/detail/TimelineFragment.kt:1606]: So I understand than when there is no typing user, there is an empty area at the bottom of the timeline, right?

File: vector/src/main/java/im/vector/app/features/home/room/detail/TimelineFragment.kt [hunk 9]
```diff
@@ -1597,6 +1601,17 @@ class TimelineFragment @Inject constructor(
...
+            state.typingUsers?.let { senders ->
+                views.typingMessageView.render(senders.take(MAX_TYPING_MESSAGE_USERS_COUNT), avatarRenderer)
```
Comment (bmarty) [inline vector/src/main/java/im/vector/app/features/home/room/detail/TimelineFragment.kt:1608]: Nit: I would rewrite to

```kotlin
state.typingUsers
    ?.take(MAX_TYPING_MESSAGE_USERS_COUNT)
    ?.let { typingUsers -> views.typingMessageView.render(typingUsers, avatarRenderer) }
```

File: vector/src/main/res/layout/fragment_timeline.xml [hunk 10]
```diff
@@ -87,8 +86,19 @@
...
+    <im.vector.app.core.ui.views.TypingMessageView
+        android:id="@+id/typingMessageView"
+        app:layout_constraintBottom_toTopOf="@id/composerLayout"
+        app:layout_constraintTop_toBottomOf="@id/timelineRecyclerView"
+        app:layout_constraintEnd_toEndOf="parent"
+        app:layout_constraintStart_toStartOf="parent"
+        android:layout_width="0dp"
+        android:paddingStart="20dp"
+        android:paddingEnd="20dp"
+        android:visibility="invisible"
```
Comment (bmarty) [inline vector/src/main/res/layout/fragment_timeline.xml:100]: Maybe make it visible in the layout editor

File: vector/src/main/res/layout/fragment_timeline.xml [hunk 11]
```diff
@@ -119,8 +129,7 @@
         android:visibility="gone"
         app:layout_constraintBottom_toBottomOf="parent"
         app:layout_constraintEnd_toEndOf="parent"
-        app:layout_constraintStart_toStartOf="parent"
-        tools:visibility="visible" />
```
Comment (bmarty) [inline vector/src/main/res/layout/fragment_timeline.xml:123]: Please revert this change. Layout editor should show all the views if possible

File: vector/src/main/res/layout/typing_message_layout.xml [hunk 12]
```diff
@@ -0,0 +1,38 @@
...
+        app:layout_constraintStart_toEndOf="@id/avatars"
+        app:layout_constraintTop_toTopOf="@id/avatars" />
+
+    <im.vector.app.core.ui.views.TypingMessageDotsView
+        android:id="@+id/viewDots"
+        android:layout_width="wrap_content"
+        android:layout_height="0dp"
+        android:layout_marginStart="8dp"
+        android:gravity="center"
+        android:visibility="visible"
```
Comment (bmarty) [inline vector/src/main/res/layout/typing_message_layout.xml:33]: This line is not necessary

File: vector/src/main/res/values/strings.xml [hunk 13]
```diff
@@ -952,6 +952,9 @@
     <string name="room_one_user_is_typing">%s is typing…</string>
     <string name="room_two_users_are_typing">%1$s &#038; %2$s are typing…</string>
     <string name="room_many_users_are_typing">%1$s &#038; %2$s &#038; others are typing…</string>
+    <!--TODO #3296 add next two strings values -->
```
Comment (bmarty) [inline vector/src/main/res/values/strings.xml:955]: This TODO is not necessary. Adding new strings will be handled automatically by Weblate

File: vector/src/main/java/im/vector/app/core/ui/views/TypingMessageAvatar.kt [hunk 14]
```diff
@@ -0,0 +1,60 @@
...
+import android.view.View
+import android.view.ViewGroup
+import android.widget.ImageView
+import android.widget.LinearLayout
+import im.vector.app.core.utils.DimensionConverter
+import im.vector.app.features.home.AvatarRenderer
+import org.matrix.android.sdk.api.session.room.sender.SenderInfo
+import org.matrix.android.sdk.api.util.toMatrixItem
+
+class TypingMessageAvatar @JvmOverloads constructor(
```
Comment (bmarty) [inline vector/src/main/java/im/vector/app/core/ui/views/TypingMessageAvatar.kt:30]: Another possible implementation could be like in the View `ReadReceiptsView`. I do not know if it is better or not though.

**Review by bmarty — COMMENTED** [review]
Some remark before we can merge this PR

**Comment by ahmed-radhouane** [comment]
@bmarty  
Any manip to avoid the : 

> 
❌ Error: 1 file has more than 2500 lines:
  ./vector/src/main/java/im/vector/app/features/home/room/detail/TimelineFragment.kt (2513 lines)

  
  which set a failing state for the Code Quality Checks ?

**Comment by bmarty** [comment]
@ahmed-radhouane you can increase the limit [here](https://github.com/vector-im/element-android/blob/main/tools/check/check_code_quality.sh#L127) (to 2800 maybe), we will have to split this class later.

File: vector/src/main/java/im/vector/app/core/ui/views/TypingMessageAvatar.kt (hunk 14 above)
Comment (ahmed-radhouane) [reply vector/src/main/java/im/vector/app/core/ui/views/TypingMessageAvatar.kt:30]: First Idea was to use this view, but too much differences with what is asked from UI team. 
But, in other way TypingMessageAvatar could be used from showing the read receipts, with some modification.

File: vector/src/main/res/values/strings.xml (hunk 13 above)
Comment (ahmed-radhouane) [reply vector/src/main/res/values/strings.xml:955]: Done. 
Thanks !

File: vector/src/main/res/layout/typing_message_layout.xml (hunk 12 above)
Comment (ahmed-radhouane) [reply vector/src/main/res/layout/typing_message_layout.xml:33]: Done. 
Thanks !

File: vector/src/main/res/layout/fragment_timeline.xml (hunk 11 above)
Comment (ahmed-radhouane) [reply vector/src/main/res/layout/fragment_timeline.xml:123]: Done. 
Thanks !

File: vector/src/main/res/layout/fragment_timeline.xml (hunk 10 above)
Comment (ahmed-radhouane) [reply vector/src/main/res/layout/fragment_timeline.xml:100]: Done. 
Thanks !

File: vector/src/main/java/im/vector/app/features/home/room/detail/TimelineFragment.kt (hunk 9 above)
Comment (ahmed-radhouane) [reply vector/src/main/java/im/vector/app/features/home/room/detail/TimelineFragment.kt:1608]: Done. 
Thanks !

File: vector/src/main/java/im/vector/app/features/home/room/detail/TimelineFragment.kt (hunk 8 above)
Comment (ahmed-radhouane) [reply vector/src/main/java/im/vector/app/features/home/room/detail/TimelineFragment.kt:1606]: Yes, it's about 20dp, otherwise we will see the jump of the time line recycler view.

File: vector/src/main/java/im/vector/app/core/ui/views/TypingMessageView.kt (hunk 7 above)
Comment (ahmed-radhouane) [reply vector/src/main/java/im/vector/app/core/ui/views/TypingMessageView.kt:48]: Done. 
Thanks !

File: vector/src/main/java/im/vector/app/core/ui/views/TypingMessageView.kt (hunk 6 above)
Comment (ahmed-radhouane) [reply vector/src/main/java/im/vector/app/core/ui/views/TypingMessageView.kt:46]: Done. 
Thanks !

File: matrix-sdk-android/src/main/java/org/matrix/android/sdk/api/session/room/sender/SenderInfo.kt (hunk 5 above)
Comment (ahmed-radhouane) [reply matrix-sdk-android/src/main/java/org/matrix/android/sdk/api/session/room/sender/SenderInfo.kt:40]: Done. 
Thanks !

File: vector/src/main/java/im/vector/app/features/home/room/detail/TimelineFragment.kt (hunk 8 above)
Comment (bmarty) [reply vector/src/main/java/im/vector/app/features/home/room/detail/TimelineFragment.kt:1610]: OK, thanks

File: vector/src/main/java/im/vector/app/features/home/room/typing/TypingHelper.kt [hunk 15]
```diff
@@ -42,4 +42,15 @@ class TypingHelper @Inject constructor(private val stringProvider: StringProvide
                         typingUsers[1].disambiguatedDisplayName)
         }
     }
+
+    fun getNotificationTypingMessage(typingUsers: List<SenderInfo>): String {
+        return when {
```
Comment (bmarty) [inline vector/src/main/java/im/vector/app/features/home/room/typing/TypingHelper.kt:47]: Do not change it, but I think we could have `when(typingUsers.size())`

Comment (bmarty) [reply vector/src/main/java/im/vector/app/features/home/room/typing/TypingHelper.kt:47]: Ah it's similar than the fun above :)

**Review by bmarty — APPROVED** [review]
Thanks for the update!
